concreteproperties/stress_strain_profile.py

```diff
@@ -457,6 +457,367 @@ def __post_init__(
+                rho_st = (
+                    self.n_confinement
+                    / self.trans_spacing
+                    * (4 * self.trans_d_b**2 * np.pi / 4 / d_s)
+                )
+                # calculate ratio of reinforcement area to core area
+                rho_cc = self.long_reinf_area / A_c
+
+                # calculate confinement effectiveness coefficient
+                exp = 2 if self.sect_type in ["circ_hoop"] else 1
+                k_e = (1 - s_dash / (2 * d_s)) ** exp / (1 - rho_cc)
+
+                # calculate tranverse confining pressures
+                # rho_b = A_vb / (self.trans_spacing * d_core)
+                f_l = k_e * rho_st * self.trans_f_y
+
+                # calculate confined concrete strength
+                f_cc = self.compressive_strength * (
+                    -1.254
+                    + 2.254 * (1 + 7.94 * f_l / self.compressive_strength) ** 0.5
+                    - 2 * f_l / self.compressive_strength
+                )
+        else:
+            # calculate unconfined concrete strength
+            f_cc = self.compressive_strength
+
+        # calculate strain associated with max confined/unconfined concrete strength
+        eps_cc = self.eps_co * (1 + 5 * (f_cc / self.compressive_strength - 1))
+
+        # calculate maximum confined/unconfined compressive strain
+        if self.conc_confined:
+            eps_c_max = min(
+                0.004
+                + self.n_steel_strain * rho_st * self.trans_f_y * self.eps_su / f_cc,
+                0.05,
+            )
+        else:
+            eps_c_max = self.eps_c_max_unconfined
+
+        # calculate secant modulus
+        E_sec = f_cc / eps_cc
+
+        if self.conc_confined:
+            self.strains = np.linspace(0, eps_c_max, self.n_points)
+            # add eps_cc point corresponding to max stress point at end
+            self.strains = np.append(self.strains, eps_cc)
+        else:
+            self.strains = np.linspace(0, min(2 * eps_cc, eps_c_max), self.n_points)
+            # add eps_cc point corresponding to max stress point at end
+            self.strains = np.append(self.strains, eps_cc)
+
+        # sort strains numerically
+        self.strains.sort()
+
+        # calculate stresses from strains
+        r = self.elastic_modulus / (self.elastic_modulus - E_sec)
+        x = self.strains / eps_cc
+        self.stresses = f_cc * x * r / (r - 1 + x**r)
+
+        # add spalling branch if specified for unconfined curve
+        if not self.conc_confined and self.conc_spalling:
+            self.strains = np.append(self.strains, self.eps_sp)
+            self.stresses = np.append(self.stresses, 0)
+
+        # calculate max tension strain based on modulus of rupture/concrete tension
+        # strength
+        eps_t = self.tensile_strength / self.elastic_modulus
+
+        if self.conc_tension:
+            # add tension stress/strain limit
+            self.strains = np.insert(self.strains, 0, -eps_t)
```

Comment:
As above, ``self.strains.insert(0, -eps_t)``.

Applies to all below.

Reply:
as per above